### ISSUE TITLE
[Compiling dfatom] `FunctionType` tied to a particular scope

### ISSUE THREAD

**Issue by czgdp1807** [lfortran/lfortran#2421]
https://github.com/czgdp1807/dftatom/tree/lf1

```
(lf) 11:34:37:~/lfortran_project/dftatom % make -f Makefile.manual  
cd src; make -f Makefile.manual
make[1]: Entering directory '/Users/czgdp1807/lfortran_project/dftatom/src'
lfortran  -c types.f90
lfortran  -c constants.f90
lfortran  -c dft_data.f90
lfortran  -c utils.f90
ASR verify pass error: ASR verify: ASR::ttype_t in ASR::FunctionType cannot be tied to a scope.
   --> utils.f90:244:1 - 251:1
    |
244 |    pure function str_real(r) result(s)
    |    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
...
    |
251 |    
    | ...^ failed here


Note: Please report unclear or confusing messages as bugs at
https://github.com/lfortran/lfortran/issues.
make[1]: *** [Makefile.manual:19: utils.o] Error 1
make[1]: Leaving directory '/Users/czgdp1807/lfortran_project/dftatom/src'
make: *** [Makefile.manual:25: all] Error 2
```

**Comment by certik** [comment]
We might need some small redesign of ASR if needed here. I think it's referencing a symbol from a type. We need to use FunctionParam, or if not possible here, let's figure out exactly the ASR situation and the right way to fix it.

**Comment by czgdp1807** [comment]
I think ASR design is correct. There might be a bug in the implementation of that ASR design I guess like some stuff got missed for conversion to `FuncParam`. I will let @Thirumalai-Shaktivel and @Shaikh-Ubaid work on it until I complete https://github.com/lfortran/lfortran/pull/2408 (I worked on it until late night yesterday and got it working locally but there are more things to be fixed which got uncovered while de-coupling function return array handling from `array_op` pass).

**Comment by Shaikh-Ubaid** [comment]
Following is a mininum reproducible example:
```fortran
% cat examples/expr2.f90 
program expr2
implicit none

integer :: p
p = 10

print *, getStr()
contains

pure function getStr() result(s)
character(len=p) :: s
s = "hi"
end function

end program
```
```bash
(lf) ubaid@ubaids-MacBook-Pro lfortran % lfortran examples/expr2.f90
ASR verify pass error: ASR verify: ASR::ttype_t in ASR::FunctionType cannot be tied to a scope.
  --> examples/expr2.f90:10:1 - 14:1
   |
10 |    pure function getStr() result(s)
   |    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
...
   |
14 |    
   | ...^ failed here


Note: Please report unclear or confusing messages as bugs at
https://github.com/lfortran/lfortran/issues.
```

**Comment by czgdp1807** [comment]
@Shaikh-Ubaid or @Thirumalai-Shaktivel Could you please create MREs for https://github.com/lfortran/lfortran/issues/2386?

**Comment by certik** [comment]
Yes, we need to support setting array sizes as parent scope variables unfortunately.

However, I don't think we use code like this in dftatom. So the above MRE is a bug, but it's not a MRE for the dftatom problem. I don't think it's good practice to be using parent scope variables for dimensions or character lengths, so I don't use it. See also https://fortran-lang.discourse.group/t/using-module-variables-as-dimensions-in-function-arguments/5742.

**Comment by Shaikh-Ubaid** [comment]
>@Shaikh-Ubaid or @Thirumalai-Shaktivel Could you please create MREs for https://github.com/lfortran/lfortran/issues/2386?

Please find https://github.com/lfortran/lfortran/issues/2386#issuecomment-1720517474.

**Comment by czgdp1807** [comment]
@Shaikh-Ubaid Please go ahead and submit a PR for this if you want.

**Comment by czgdp1807** [comment]
> However, I don't think we use code like this in dftatom. So the above MRE is a bug, but it's not a MRE for the dftatom problem. I don't think it's good practice to be using parent scope variables for dimensions or character lengths, so I don't use it.

Ah I see. So the MRE itself is not correct, so here's what needs to be done,

- [ ] Create a correct MRE.
- [ ] Then fix it.
- [ ] Test the fix with `dftatom`
- [ ] Merge the PR with the fix.

@Shaikh-Ubaid Please keep us updated on it. As soon as I am done with #2408, I will help you.

**Comment by Shaikh-Ubaid** [comment]
>So I suggest we don't worry about this for dftatom, since it does not use this feature. I recommend to first fix issues in dftatom, then worry about other bugs.

@czgdp1807 I think this issue is not to be solved. As Ondrej Sir suggested, we should not worry about this for `dfatom`.

Also, I think this is a bug in LFortran with the correct MRE, but the bug is not related to `dfatom` as Ondrej Sir said.

**Comment by czgdp1807** [comment]
Well what needs to be fixed is the issue described in https://github.com/lfortran/lfortran/issues/2421#issue-1895775561. Either workaround it in `dftatom` codebase or let's fix AST to ASR or ASR to ASR passes (wherever the bug is in LFortran).

For adding a workaround you can do the following,

1. Create a new branch (like `lf2`) after cloning my fork, https://github.com/czgdp1807/dftatom.
2. Add your workaround so that the compilation progresses.
3. Push your branch into my fork.

I will give you write access to my fork in a moment.

For fixing the issue I described the steps in https://github.com/lfortran/lfortran/issues/2421#issuecomment-1721195706.

**Comment by Shaikh-Ubaid** [comment]
@czgdp1807 please see https://github.com/lfortran/lfortran/issues/2386#issue-1886204065. This code that gives the error in https://github.com/lfortran/lfortran/issues/2421#issue-1895775561 is itself removed.

Thus, I think (and as Ondrej Sir shared) this does not seem to be a priority for the moment.

**Comment by czgdp1807** [comment]
Sure. Then let's fix #2386 first or add an workaround for it. I assigned you that. Probably, this issue can also be workaround. If we are not using `str_real` and other such functions then let's just ignore this issue for now and keep it open.
